File: Data Structure/Stack/155.min_stack.js
/**
 * initialize your data structure here.
 */
var MinStack = function() {
    this.minStack = [];
    this.stack = [];
};

/**
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
   this.stack.push(x);
   if (this.minStack.length === 0) {
       this.minStack.push(x);
   } else {
       this.minStack.push(Math.min(x, this.minStack[this.minStack.length - 1]));
   }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
   this.stack.pop();
   this.minStack.pop();
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return this.stack[this.stack.length - 1];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return this.minStack[this.minStack.length - 1];
};

File: Java/155.min_stack.java
// 用2个STACK,一个存储最小值，一个存储所有值
// MIN STACK里存储的是对于当前ELEMENT的最小值
// 需要记忆的算法
class MinStack {
    Stack<Integer> stack;
    Stack<Integer> minStack;

    public MinStack() {
        stack = new Stack<>();
        minStack = new Stack<>();
    }

    public void push(int x) {
        stack.push(x);
        if (minStack.isEmpty()) {
            minStack.push(x);
        } else {
            minStack.push(Math.min(x, minStack.peek()));
        }
    }

    public int pop() {
        minStack.pop();
        return stack.pop();
    }

    public int top() {
        return stack.peek();
    }
    public int getMin() {
        return minStack.peek();
    }
}



// Discuss 神仙解法，跟上面的解法效率一样的
class MinStack {
    Stack<Integer> stack;
    int min = Integer.MAX_VAL;
    /** initialize your data structure here. */
    public MinStack() {
        stack = new Stack<>();
    }

    public void push(int x) {
        if (x <= min) {
            stack.push(min);
            min = x;
        }
    }

    public void pop() {
        if (stack.peek() == min) min = stack.pop();
    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return min;
    }
}